### PR TITLE
ENH: add export callback

### PR DESCRIPTION
**Pull request by CJ-Wright**
Add callback for exporting data to a new databroker

### CODE REVIEW

**Comment by CJ-Wright** [comment]
I'm not exactly certain what the error is here, it seems to be in the sqlite layer. Is it possible that there is an error on the databroker side?

**Comment by tacaswell** [comment]
Can you reproduce the failure locally?

**Comment by CJ-Wright** [comment]
Yes, its the same locally (I put this up so I could check it was just  a "my janky machine is janky")

**Comment by CJ-Wright** [comment]
@tacaswell I think this is ready for another round of review!

File: bluesky/callbacks/broker.py [hunk 1]
```diff
@@ -301,3 +301,37 @@ def stop(self, doc):
...
+        """
+        self.new_db = new_db
+        self.old_db = old_db
+        self.copy_kwargs = copy_kwargs
+        self.new_root = new_root
+
+        self.start_uid = None
+
+    def start(self, doc):
+        # This is run on start just in case we miss a stop
```
Comment (danielballan) [inline bluesky/callbacks/broker.py]: By what mechanism are we expecting to miss a stop?

Comment (CJ-Wright) [reply bluesky/callbacks/broker.py]: I think I thought that halt wouldn't put out a stop. I can remove it though.

Comment (danielballan) [reply bluesky/callbacks/broker.py]: That is correct, but I wouldn't both designing around that case.

**Comment by CJ-Wright** [comment]
I've seen the travis error on my local (als py35). If I throw in a print statement the error changes to something about joining string and path objects not supported.

**Comment by danielballan** [comment]
Would you mind pasting the Traceback here out of Travis?

**Comment by CJ-Wright** [comment]
```python
RE = <bluesky.run_engine.RunEngine object at 0x7fcb7df52438>
db = <databroker._core.Broker object at 0x7fcbb97b5668>
db2 = <databroker._core.Broker object at 0x7fcb65fd9d30>
tmpdir = local('/tmp/pytest-of-travis/pytest-0/test_exporter0')
    def test_exporter(RE, db, db2, tmpdir):
        RE.subscribe(db.insert)
        db.reg.register_handler('RWFS_NPY', NumpySeqHandler)
        det = SynSignalWithRegistry(lambda: np.array(np.ones((10, 10))),
                                    name='det')
        exporter = Exporter(db2, db, tmpdir)
        RE.subscribe(exporter)
>       RE(count([det]))
bluesky/tests/test_callbacks.py:462: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
bluesky/run_engine.py:744: in __call__
    raise exc
/opt/python/3.5.5/lib/python3.5/asyncio/tasks.py:240: in _step
    result = coro.send(None)
bluesky/run_engine.py:1253: in _run
    raise err
bluesky/run_engine.py:1140: in _run
    msg = self._plan_stack[-1].send(resp)
bluesky/plans.py:67: in count
    return (yield from inner_count())
bluesky/utils.py:1013: in dec_inner
    return (yield from plan)
bluesky/preprocessors.py:939: in stage_wrapper
    return (yield from finalize_wrapper(inner(), unstage_devices()))
bluesky/preprocessors.py:500: in finalize_wrapper
    ret = yield from plan
bluesky/preprocessors.py:937: in inner
    return (yield from plan)
bluesky/utils.py:1013: in dec_inner
    return (yield from plan)
bluesky/preprocessors.py:326: in run_wrapper
    else_plan=close_run)
bluesky/preprocessors.py:576: in contingency_wrapper
    yield from else_plan()
bluesky/plan_stubs.py:733: in close_run
    return (yield Msg('close_run', exit_status=exit_status, reason=reason))
bluesky/run_engine.py:1197: in _run
    new_response = yield from coro(msg)
bluesky/run_engine.py:1422: in _close_run
    yield from self.emit(DocumentNames.stop, doc)
/opt/python/3.5.5/lib/python3.5/asyncio/coroutines.py:210: in coro
    res = func(*args, **kw)
bluesky/run_engine.py:2311: in emit
    self.dispatcher.process(name, doc)
bluesky/run_engine.py:2331: in process
    exceptions = self.cb_registry.process(name, name.name, doc)
bluesky/utils.py:336: in process
    func(*args, **kwargs)
bluesky/utils.py:417: in __call__
    return mtd(*args, **kwargs)
bluesky/callbacks/core.py:26: in __call__
    return getattr(self, name)(doc)
bluesky/callbacks/broker.py:327: in stop
    self.copy_kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <databroker._core.Broker object at 0x7fcbb97b5668>
headers = [Header(db=<databroker._core.Broker object at 0x7fcbb97b5668>, start={'detectors': ['det'], 'plan_type': 'generator', ...}, 'plan_name': 'count', 'time': 1530627331.8140976, 'num_intervals': 0}, stop={'_name': 'Run Stop'}, ext=namespace())]
db = <databroker._core.Broker object at 0x7fcb65fd9d30>
new_root = local('/tmp/pytest-of-travis/pytest-0/test_exporter0')
copy_kwargs = {}
    def export(self, headers, db, new_root=None, copy_kwargs=None):
        """
            Export a list of headers.
    
            Parameters
            ----------
            headers : databroker.header
                one or more headers that are going to be exported
            db : databroker.Broker
                an instance of databroker.Broker class that will be the target to
                export info
            new_root : str
                optional. root directory of files that are going to
                be exported
            copy_kwargs : dict or None
                passed through to the ``copy_files`` method on Registry;
                None by default
    
            Returns
            ------
            file_pairs : list
                list of (old_file_path, new_file_path) pairs generated by
                ``copy_files`` method on Registry.
            """
        if copy_kwargs is None:
            copy_kwargs = {}
        try:
            headers.items()
        except AttributeError:
            pass
        else:
            headers = [headers]
        file_pairs = []
        for header in headers:
            # insert mds
            db.mds.insert_run_start(**_sanitize(header['start']))
            events = self.get_events(header)
            for descriptor in header['descriptors']:
                db.mds.insert_descriptor(**_sanitize(descriptor))
                for event in events:
                    db.mds.insert_event(**_sanitize(event))
>           db.mds.insert_run_stop(**_sanitize(header['stop']))
E           TypeError: insert_run_stop() missing 3 required positional arguments: 'run_start', 'time', and 'uid'
../../../virtualenv/python3.5.5/lib/python3.5/site-packages/databroker/_core.py:1864: TypeError
```

**Comment by CJ-Wright** [comment]
I'm going to close this in favor of interacting with the resource and datum documents directly.

**Pull request closed without merging**